transport/udp/udp.go

```diff
@@ -0,0 +1,390 @@
+	// less (not considering UDPv6 jumbograms, RFC 2675).
+	// These are theoretical limits, since in practice the maximum datagram size
+	// is determined by the MTU (typically 1500 .. ~9000 bytes).
+	// Hence a buffer size of 64k should suffice for 'regular' networks.
+	udpMaxBuf = 1 << 16
+)
+
+// udpEndpoint represents attributes and methods common to udp Dialers/Listeners
+type udpEndpoint struct {
+	addr  *net.UDPAddr           // Dialer: remote address to connect to; Listener: local address
+	opts  map[string]interface{} // Dialer or Listener options
+	proto mangos.Protocol        // Local protocol
+}
+
+func (e *udpEndpoint) SetOption(name string, value interface{}) error {
+	// FIXME: nothing at the moment
+	return mangos.ErrBadOption
+}
+
+func (e *udpEndpoint) GetOption(name string) (interface{}, error) {
+	v, ok := e.opts[name]
+	if !ok {
+		return nil, mangos.ErrBadOption
+	}
+	return v, nil
+}
+
+func (e *udpEndpoint) LocalProtocol() uint16 {
+	return e.proto.Number()
+}
+
+func (e *udpEndpoint) RemoteProtocol() uint16 {
+	return e.proto.PeerNumber()
+}
+
+// extractMsgFromPacket validates the SP header in @pkt and extracts it into a Message.
+// FIXME: this could be refactored into conn.go
+func (e *udpEndpoint) extractMsgFromPacket(pkt []byte) (*mangos.Message, error) {
+	var h connHeader
+	var r = bytes.NewBuffer(pkt)
+
+	if err := binary.Read(r, binary.BigEndian, &h); err != nil {
+		return nil, err
+	}
+
+	if h.Zero != 0 || h.S != 'S' || h.P != 'P' || h.Rsvd != 0 {
+		return nil, mangos.ErrBadHeader
+	}
+
+	// The only version number we support at present is "0", at offset 3.
+	if h.Version != 0 {
+		return nil, mangos.ErrBadVersion
+	}
+
+	// The protocol number lives as 16-bits (big-endian) at offset 4.
+	if h.Proto != e.proto.PeerNumber() {
+		return nil, mangos.ErrBadProto
+	}
+
+	msg := mangos.NewMessage(r.Len())
+	msg.Body = r.Bytes()
+
+	return msg, nil
+}
+
+// udpPipe implements the Pipe interface on top of a UDP datagram connection.
+type udpPipe struct {
+	udpEndpoint
+	*net.UDPConn      // bidirectional or half-open UDP "connection"
+	open         bool // whether the net.UDPConn is in a usable state
+	props        map[string]interface{}
+}
+
+func (p *udpPipe) GetProp(name string) (interface{}, error) {
+	if v, ok := p.props[name]; ok {
+		return v, nil
+	}
+	return nil, mangos.ErrBadProperty
+}
+
+func (p *udpPipe) Close() error {
+	if !p.open {
+		return nil
+	}
+	p.open = false
+	return p.UDPConn.Close()
+}
+
+func (u *udpPipe) IsOpen() bool {
+	return u.open
+}
+
+func (p *udpPipe) sendHelper(m *mangos.Message, writePkt func(b []byte) (int, error)) error {
+	if !m.Expired() {
+		var b bytes.Buffer
+		var h = connHeader{S: 'S', P: 'P', Proto: p.proto.Number()}
+
+		if err := binary.Write(&b, binary.BigEndian, &h); err != nil {
+			return err
+		}
+		if _, err := b.Write(m.Header); err != nil {
+			return err
+		}
+		if _, err := b.Write(m.Body); err != nil {
```

Comment:
Probably we should check for the case of a message that is too big.  We need to return the error case of a message that won't fit into a UDP message to the caller somehow. 

This also underscores our need for statistics...